### PR TITLE
HTTPProxy: add healthCheck port config

### PR DESCRIPTION
**Pull request by yangyy93**
Add healthCheck port for httproute and tcproute

fix https://github.com/projectcontour/contour/issues/4655
Signed-off-by: yy [yang.yang@daocloud.io]

### CODE REVIEW

File: apis/projectcontour/v1/httpproxy.go [hunk 1]
```diff
@@ -786,6 +786,12 @@ type Service struct {
 	// +kubebuilder:validation:ExclusiveMinimum=false
 	// +kubebuilder:validation:ExclusiveMaximum=true
 	Port int `json:"port"`
+	// HealthPort is the port for this service healthcheck.
+	// if the field were not specified, it is the same as Port.
+	//
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65536
```
Comment (sunjayBhatia) [inline apis/projectcontour/v1/httpproxy.go:793]: should add the `+optional` annotation here as well

File: apis/projectcontour/v1/httpproxy.go [hunk 2]
```diff
@@ -786,6 +786,12 @@ type Service struct {
 	// +kubebuilder:validation:ExclusiveMinimum=false
 	// +kubebuilder:validation:ExclusiveMaximum=true
 	Port int `json:"port"`
+	// HealthPort is the port for this service healthcheck.
+	// if the field were not specified, it is the same as Port.
```
Comment (sunjayBhatia) [inline apis/projectcontour/v1/httpproxy.go:790]:
```suggestion
	// If not specified, Port is used for service healthchecks.
```

File: changelogs/unreleased/4761-yangyy93-minor.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+## Add health check port for HTTP health check and TCP health check
+
+HTTPProxy.Route.Service and HTTPProxy.TCPProxy.Service now has a `HealthPort` which specifies a health check port that is different from the routing port.If the field was not specified,it is the same as `Port`.
```
Comment (sunjayBhatia) [inline changelogs/unreleased/4761-yangyy93-minor.md:3]:
```suggestion
HTTPProxy.Route.Service and HTTPProxy.TCPProxy.Service now has a `HealthPort` field which specifies a health check port that is different from the routing port. If not specified, the service `Port` field is used for healthchecking.
```

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks for the PR @yangyy93!

Please also add some documentation to https://github.com/projectcontour/contour/blob/main/site/content/docs/main/config/health-checks.md on this new feature.

File: apis/projectcontour/v1/httpproxy.go (hunk 2 above)
Comment (yangyy93) [reply apis/projectcontour/v1/httpproxy.go]: thanks for your comment

File: apis/projectcontour/v1/httpproxy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yangyy93) [reply apis/projectcontour/v1/httpproxy.go]: thanks for your comment，the relevant documents have been supplemented

**Comment by yangyy93** [comment]
Thank you for your comments, the relevant documents and comments have been revised.
@skriss @sunjayBhatia

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks for the updates @yangyy93, two more small comments.

File: internal/dag/accessors.go [hunk 4]
```diff
@@ -27,12 +27,20 @@ import (
...
 		return nil, err
 	}
 
+	if healthPort.IntVal == 0 && healthPort.StrVal == "" {
+		healthPort = port
+	}
+	_, healthSvcPort, err := cache.LookupService(meta, healthPort)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (yangyy93) [reply internal/dag/accessors.go]: Thanks for your comments

**Review by skriss — APPROVED** [review]
LGTM, will leave for @sunjayBhatia to take another look

File: internal/envoy/v3/endpoint_test.go [hunk 5]
```diff
@@ -32,6 +32,19 @@ func TestLBEndpoint(t *testing.T) {
 	protobuf.ExpectEqual(t, want, got)
 }
 
+func TestHealthCheckLBEndpoint(t *testing.T) {
+	got := HealthCheckLBEndpoint(SocketAddress("microsoft.com", 81), 8998)
+	want := &envoy_endpoint_v3.LbEndpoint{
+		HostIdentifier: &envoy_endpoint_v3.LbEndpoint_Endpoint{
+			Endpoint: &envoy_endpoint_v3.Endpoint{
+				Address:           SocketAddress("microsoft.com", 81),
+				HealthCheckConfig: HealthCheckConfig(8998),
```
Comment (sunjayBhatia) [inline internal/envoy/v3/endpoint_test.go:41]: would prefer this not use a helper to fill in this field but rather explicitly write out the object we expect, just to ensure we're not using something under test to generate the expected value

File: internal/dag/accessors_test.go [hunk 6]
```diff
@@ -41,6 +41,29 @@ func TestBuilderLookupService(t *testing.T) {
...
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				{
+					Name:       "http",
+					Protocol:   "TCP",
+					Port:       8080,
+					TargetPort: intstr.FromInt(8080),
+				},
+				{
+					Name:       "heath",
```
Comment (sunjayBhatia) [inline internal/dag/accessors_test.go:58]: nit
```suggestion
					Name:       "health",
```

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]
logic looks good, but some changes requested on the structure of the endpointstranslator tests

**Comment by yangyy93** [comment]
> 逻辑看起来不错，但需要对端点翻译器测试的结构进行一些更改

do you have any other suggestions？

**Comment by yangyy93** [comment]
> logic looks good, but some changes requested on the structure of the endpointstranslator tests

do you have any other suggestions？ @sunjayBhatia

**Comment by skriss** [comment]
@sunjayBhatia I think this one's just awaiting another look from you

**Review by sunjayBhatia — APPROVED** [review]
lgtm, I may make a follow up PR since that is a little easier to fix some test things

**Comment by sunjayBhatia** [comment]
follow up: https://github.com/projectcontour/contour/pull/4828

**Pull request merged**